pkg/sanity/controller.go

```diff
@@ -1065,90 +1072,14 @@ var _ = DescribeSanity("Controller Service [Controller Server]", func(sc *TestCo
+		})
 
-			By("getting node information")
-			ni, err := n.NodeGetInfo(
-				context.Background(),
-				&csi.NodeGetInfoRequest{})
-			Expect(err).NotTo(HaveOccurred())
-			Expect(ni).NotTo(BeNil())
-			Expect(ni.GetNodeId()).NotTo(BeEmpty())
-
-			var accReqs *csi.TopologyRequirement
-			if ni.AccessibleTopology != nil {
-				// Topology requirements are honored if provided by the driver
-				accReqs = &csi.TopologyRequirement{
-					Requisite: []*csi.Topology{ni.AccessibleTopology},
-				}
-			}
-
-			// Create Volume First
-			By("creating a single node writer volume")
-			name := UniqueString("sanity-controller-publish")
-
-			vol, err := c.CreateVolume(
-				context.Background(),
-				&csi.CreateVolumeRequest{
-					Name: name,
-					VolumeCapabilities: []*csi.VolumeCapability{
-						TestVolumeCapabilityWithAccessType(sc, csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER),
-					},
-					Secrets:                   sc.Secrets.CreateVolumeSecret,
-					Parameters:                sc.Config.TestVolumeParameters,
-					AccessibilityRequirements: accReqs,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(vol).NotTo(BeNil())
-			Expect(vol.GetVolume()).NotTo(BeNil())
-			Expect(vol.GetVolume().GetVolumeId()).NotTo(BeEmpty())
-			cl.RegisterVolume(name, VolumeInfo{VolumeID: vol.GetVolume().GetVolumeId()})
-
-			// ControllerPublishVolume
-			By("calling controllerpublish on that volume")
-
-			conpubvol, err := c.ControllerPublishVolume(
-				context.Background(),
-				&csi.ControllerPublishVolumeRequest{
-					VolumeId:         vol.GetVolume().GetVolumeId(),
-					NodeId:           ni.GetNodeId(),
-					VolumeCapability: TestVolumeCapabilityWithAccessType(sc, csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER),
-					Readonly:         false,
-					Secrets:          sc.Secrets.ControllerPublishVolumeSecret,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			cl.RegisterVolume(name, VolumeInfo{VolumeID: vol.GetVolume().GetVolumeId(), NodeID: ni.GetNodeId()})
-			Expect(conpubvol).NotTo(BeNil())
-
-			By("cleaning up unpublishing the volume")
-
-			conunpubvol, err := c.ControllerUnpublishVolume(
-				context.Background(),
-				&csi.ControllerUnpublishVolumeRequest{
-					VolumeId: vol.GetVolume().GetVolumeId(),
-					// NodeID is optional in ControllerUnpublishVolume
-					NodeId:  ni.GetNodeId(),
-					Secrets: sc.Secrets.ControllerUnpublishVolumeSecret,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			Expect(conunpubvol).NotTo(BeNil())
-
-			By("cleaning up deleting the volume")
-
-			_, err = c.DeleteVolume(
-				context.Background(),
-				&csi.DeleteVolumeRequest{
-					VolumeId: vol.GetVolume().GetVolumeId(),
-					Secrets:  sc.Secrets.DeleteVolumeSecret,
-				},
-			)
-			Expect(err).NotTo(HaveOccurred())
-			cl.UnregisterVolume(name)
+		It("should return appropriate values (no optional values added), with idempotency", func() {
+			By("lifecycle repeated")
+			VolumeLifecycle(n, c, sc, cl, idempotentcount)
```

Comment:
You are right, "should return appropriate values (no optional values added)" doesn't make a lot of sense. I would go one step further than merely renaming it: this and the idempotency variant are basically a full test for "Controller Service", not just for "Controller Service ControllerPublishVolume", which is where it is currently defined.

Can you move it out of `Describe("ControllerPublishVolume", func() {` and instead use `Describe("works", ...` for the simple test and `Describe("is idempotent", ...` for the one with repetition?

Reply:
Updated the comment: it now uses exactly the same description as the corresponding node tests.

Reply:
pushed new version where I tried to follow these ideas.